lms/djangoapps/onboarding_survey/history.py

```diff
@@ -0,0 +1,186 @@
+
+    for interested_communities_choice in interests_survey.interested_communities.all():
+        setattr(history, many_to_many_fields_mapping[interested_communities_choice.community_type], True)
+
+
+def add_organization_survey(history, user):
+    """
+    Function to add organization survey in history
+
+    Arguments:
+        history(History): History table model
+        user(User): Django user model instance
+    """
+
+    # The fields to be added in the history table from organization survey
+    fields = ['role_in_org', 'partner_network']
+    # Some of the fields in history table have different corresponding names in organization survey.
+    history_and_org_survey_field_mappings = {
+        'org_sector': 'sector',
+        'org_level_of_operation': 'level_of_operation',
+        'org_focus_area': 'focus_area',
+        'org_total_employees': 'total_employees',
+        'org_total_volunteers': 'total_volunteers',
+        'org_start_month_year': 'start_month_year',
+        'org_country': 'country',
+        'org_city': 'city',
+        'org_url': 'url',
+        'org_founding_year': 'founding_year',
+        'org_total_annual_clients_or_beneficiary': 'total_annual_clients_or_beneficiary',
+        'org_total_annual_revenue_for_last_fiscal_year': 'total_annual_revenue_for_last_fiscal_year',
+    }
+
+    organization_survey = user.organization_survey
+    add_fields(fields, organization_survey, history, history_and_org_survey_field_mappings)
+
+
+def add_extended_registration(history, user):
+    """
+    Function to add extended profile fields in history
+
+    Arguments:
+        history(History): History table model
+        user(User): Django user model instance
+    """
+
+    # The fields to be added in the history table from extended profile
+    fields = ['first_name', 'last_name', 'organization', 'is_poc', 'is_currently_employed', 'org_admin_email']
+
+    extended_profile = user.extended_profile
+    add_fields(fields, extended_profile, history)
+
+
+def populate_history(user):
+    """
+    populate user history.
+
+    Arguments:
+        user(User): Django user model instance
+
+    Returns
+        history: Newly created History model instance
+    """
+    history = History()
+    add_user_info_survey(history, user)
+    add_interests_survey(history, user)
+    add_organization_survey(history, user)
+    add_extended_registration(history, user)
+    history.user = user
+    return history
+
+
+def save_history(history, user):
+    """
+    Save user history.
+
+    It will update the previous(if any) entry of history for the same user
+    by added value in end_date column. It then, will create a new entry in the
+    history table.
+
+    Arguments:
+        history(History): History table model
+        user(User): Django user model instance
+    """
+    most_recent_history = History.objects.filter(user=user, end_date=None).first()
+    if most_recent_history:
```

Comment:
What is the reason for updating `end_date` here?

Reply:
We want to keep track of time span in which a particular record was active. For example, when a user fills all surveys for the first time, we create a row in the history table with end_date null. Now, if user updates his/her info, we go back to our history table, find out which row corresponding to that user has empty end_date, update this record's end_date with the current timestamp, save the record and then create a new record corresponding to the changed values in surveys and set this record's end_date to null.